docs/generate-and-template.md

```diff
@@ -0,0 +1,294 @@
+output to a file with the `-o` or `--out` flags:
+
+```
+$ helm template -o out.yaml -d values.toml example.yaml
+```
+
+This final form is the one used most frequently by generators.
+
+## Generators
+
+Helm provides a command called `helm generate`. This function operates
+on charts within your workspace (e.g. charts that you have fetched or
+installed).
+
+The generator's job is to _scan the files inside of a chart looking for
+`helm:generate` headers_.
+
+Files that begin with the appropriate `helm:generate` header will
+trigger a generation run.
+
+### The `helm:generate` Header
+
+The `helm:generate` header must be the first line of a file (any file),
+and must exactly follow one of the three formats described below:
+
+```
+#helm:generate CMD [ARGS]
+//helm:generate CMD [ARGS]
+/*helm:generate CMD [ARGS]*/
+```
+
+The generate header is space-sensitive and case-sensitive. Lines MUST
+begin with one of the above comment sequences, and MUST have a lowercase
+`helm:generate` string.
+
+`CMD` is the command that Helm will run, and the optional `[ARGS]` are
+arguments passed to the `CMD`.
+
+For example, we could embed a simple `sed` script inside of a generate
+header:
+
+```
+#helm:generate sed -i -e s|ubuntu-debootstrap|fluffy-bunny| my/pod.yaml
+```
+
+It is important to note that generate commands are _not run in a shell_.
+However, environment variables are expanded.
+
+Along with any existing environment variables, the following variables
+are specially defined:
+
+- HELM_HOME: The Helm home directory
+- HELM_DEFAULT_REPO: The repository alias for the default repository.
+- HELM_GENERATE_FILE: The present file's name
+- HELM_GENERATE_DIR: The absolute path to the chart directory of the present chart
+
+These are available both from the invocation of the `CMD`, and from
+inside any generator command itself.
+
+As an example, consider a generation header like this, inside of a file
+named `example.js`:
+
+```
+//helm:generate echo $HELM_GENERATE_FILE
+```
+
+This will be expanded to the command `echo /path/to/example.json`.
+
+### Writing A Custom Generator
+
+A generator is any tool that is executable within your environment. When
+`helm generate` runs, it looks for the generator in two places:
+
+- First, if the generator is an absolute path (any path that begins with
+  `/`), it will execute it on that path.
+- Otherwise it will search for `CMD` on `$PATH`.
+
+To execute a script relative to the chart's path, you can use
+the `$HELM_GENERATE_DIR` variable:
+
+```
+/*helm:generate $HELM_GENERATE_DIR/myscript.sh */
+```
+
+Note that the above variable is the absolute path to the chart
+directory, not to the file that `helm generate` found.
+
+The following guidelines should be kept in mind when writing Helm
+generator plugins:
+
+- All of the named environment variables are available to the plugin.
+- The plugin does _not_ execute in a shell. Thus you cannot expect that
+  implementors will be able to easily redirect STDOUT or STDERR
```

Comment:
It redirects the stdout/stderr to the current STDOUT/STDERR that the app executes in. So, for example, `helm generate | grep "bar"` will capture any case where a plugin emitted `bar`.

The main issue I was trying to convey here is that you can't do:

```
helm:generate foo > bar
```

And that's really because `foo > bar` does not execute in a shell.